File: python-algorithm/ds_3_hash/L409.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/1 21:26
# @Author  : Heng Li
# @File    : L409.py
# @Software: PyCharm

# 409. 最长回文串

# 乍一看以为要考察其它算法，用哈希就可以解决了

class Solution:
    def longestPalindrome(self, s: str) -> int:
        d = {}
        res = 1
        for c in s:
            if c in d:
                d[c] += 1
            else:
                d[c] = 1
        for k,v in d.items():
            if v%2 == 0:
                res += v
            else:
                res += v-1
        return len(s) if res>len(s) else res

File: ds_3_hash/L409.cpp

#include <iostream>
#include <unordered_map>

using namespace std;
class Solution {
public:
    int longestPalindrome(string s) {
        int res = 0 ;
        int flag = 0;
        unordered_map<char,int> m;
        for(auto c:s)
            m[c]++;
        for(auto p:m){
            if(p.second%2 == 0)
                res += p.second;
            else{
                res += p.second-1;
                flag = 1;
            }
        }
        return flag == 0? res:res+1;
    }
};